danceradvent/public/articles/2023/20-dancing-with-htmx.pod

```diff
@@ -0,0 +1,333 @@
+
+=over 4
+
+=item hx-get
+
+This is the HTML verb to use in the request. We make a C<GET> request to
+C</edit/ID> and it returns HTML that replaces the C<E<lt>tdE<gt>> containing
+the name with a simple form for editing the name.
+
+=item hx-target
+
+Once again, the HTML element that is clicked (the edit icon) is not the one
+that needs to be replaced. We've given the correct C<E<lt>tdE<gt>> element the
+ID C<edit-ID>, so we use that as the target.
+
+=back
+
+We now need to implement the C<get '/edit/:id'> route in our Dancer app. It
+needs to return the edit form, so it looks like this:
+
+    get '/edit/:id' => sub {
+      my $id = route_parameters->get('id');
+
+      my $sch = HTMXmas::Schema->connect(...);
+      my $rs  = $sch->resultset('Child');
+
+      my $row = $rs->find($id);
+
+      return template 'edit_name.tt', {
+        row => $row,
+      };
+    };
+
+And the C<edit_name.tt> template looks like this:
+
+    <form>
+      <input class="form-control" id="name" name="name" value="<% row.name %>">
+      <i hx-post="/update/<% row.id %>" hx-params="<% col %>" hx-ext="json-enc"
+         hx-target="#edit-<% row.id %>" title="save" class="bi bi-check-square
+         text-success" style="font-size: 1.25rem"></i>
+      <i hx-get="/reset/<% row.id %>" hx-target="#edit-<% row.id %>"
+         title="cancel" class="bi bi-x-square text-danger"
+         style="font-size: 1.25rem"></i>
+    </form>
+
+So what we're doing here is replacing some HTML with other HTML that includes
+C<hx-*> attributes. Editing the name is a two-stage process. The first stage
+displays the form for editing the name and the second stage saves any changes
+to the database.
+
+Notice that we've added two icons to the edit form. There's a check mark to
+save the changes to the database and an X that cancels any changes. After
+pressing either of those, the form should be replaced with the original
+C<E<lt>tdE<gt>> - so that's the HTML that the routes should return.
+
+We understand most of the C<hx-*> attributes that are used here, so I won't
+go into them again. There's just one new one to explain.
+
+=over 4
+
+=item hx-ext
+
+There are a number of extensions to the basic HTMX features. By default,
+a C<POST> request like we're using for the update here would use the standard
+C<application/x-www-form-urlencoded> encoding. But in this case, Pixie decided
+to use an extension (called "json-enc") which, instead, encodes the data as
+JSON. She thinks that makes it a little easier to deal with the data at the
+in the Dancer app.
+
+=back
+
+Which brings us neatly back to the Dancer app. We need to add two new routes
+to handle these icon - C<post '/update/:id/'> will save the data to the
+database (and remove the form from the table) and C<get '/reset/:id'> will
+undo any changes and remove the form. They look like this:
+
+    post '/update/:id' => sub {
+      my $id = route_parameters->get('id');
+
+      my $object = from_json($request->body);
```

Comment:
Drop the leading $ from request here since it's a keyword.

Reply:
Ah yes. That was simplified from a longer version where that line of code was in a subroutine that took the request as a parameter. Thanks.